### DISCUSSION TITLE
document title isn't changing according

### DISCUSSION THREAD

**Discussion by SH20RAJ** [vercel/next.js#63719] [Help]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
The title isn't changing according. I have used api to fetch the data and here is my website 

https://dev-art.vercel.app/

and here is the code link :- https://github.com/SH20RAJ/devart

When I click on any post the post is opening the content is loading well without reload but the document title is changing to the title of some other post, sometimes it also works fine but maximum times it doesn't work, after reloading again and again the title changes every time that not supposed to be changed and after some reload it becomes correct 💯 and expected title, this thing also happens in the view-source of the website.

The meta Twitter card is working fine everytime no problem it's a title issue.

See my codes on GitHub :- https://github.com/SH20RAJ/devart/tree/7d72df6916d01866a2efc202d893bfba3147482d (commit URL in case of I change the codes later in try of fixing the bug.)



You should add Next Head in app router also.... (Extra recommendation)
```

### Which example does this report relate to?

https://dev-art.vercel.app/

### How are you deploying your application? (if relevant)

See on my GitHub

### Describe the Bug

The title isn't changing according. I have used api to fetch the data and here is my website 

https://dev-art.vercel.app/

and here is the code link :- https://github.com/SH20RAJ/devart

When I click on any post the post is opening the content is loading well without reload but the document title is changing to the title of some other post, sometimes it also works fine but maximum times it doesn't work, after reloading again and again the title changes every time that not supposed to be changed and after some reload it becomes correct 💯 and expected title, this thing also happens in the view-source of the website.

The meta Twitter card is working fine everytime no problem it's a title issue.

See my codes on GitHub :- https://github.com/SH20RAJ/devart/tree/7d72df6916d01866a2efc202d893bfba3147482d (commit URL in case of I change the codes later in try of fixing the bug.)

You should add Next Head in app router also.... (Extra recommendation)

### Expected Behavior

The title isn't changing according. I have used api to fetch the data and here is my website 

https://dev-art.vercel.app/

and here is the code link :- https://github.com/SH20RAJ/devart

When I click on any post the post is opening the content is loading well without reload but the document title is changing to the title of some other post, sometimes it also works fine but maximum times it doesn't work, after reloading again and again the title changes every time that not supposed to be changed and after some reload it becomes correct 💯 and expected title, this thing also happens in the view-source of the website.

The meta Twitter card is working fine everytime no problem it's a title issue.

See my codes on GitHub :- https://github.com/SH20RAJ/devart/tree/7d72df6916d01866a2efc202d893bfba3147482d (commit URL in case of I change the codes later in try of fixing the bug.)

You should add Next Head in app router also.... (Extra recommendation)

### To Reproduce

The title isn't changing according. I have used api to fetch the data and here is my website 

https://dev-art.vercel.app/

and here is the code link :- https://github.com/SH20RAJ/devart

When I click on any post the post is opening the content is loading well without reload but the document title is changing to the title of some other post, sometimes it also works fine but maximum times it doesn't work, after reloading again and again the title changes every time that not supposed to be changed and after some reload it becomes correct 💯 and expected title, this thing also happens in the view-source of the website.

The meta Twitter card is working fine everytime no problem it's a title issue.

See my codes on GitHub :- https://github.com/SH20RAJ/devart/tree/7d72df6916d01866a2efc202d893bfba3147482d (commit URL in case of I change the codes later in try of fixing the bug.)

You should add Next Head in app router also.... (Extra recommendation)

**Comment by devjiwonchoi** [comment]
Hi, I'd recommend you to first try to use the combination of `generateMetadata` and proper params from dynamic routes:

```ts
export default function Page({ params }: { params: { slug: string } }) {
  // fetch dev.to with correct slug
}

export async function generateMetadata({ params }) {
  // fetch dev.to with correct slug
  return {
    title: '...',
  }
}
```

ref: https://nextjs.org/docs/app/api-reference/functions/generate-metadata

**Comment by SH20RAJ** [comment]
I have solved the issue using layout.js:
See Here :- https://github.com/SH20RAJ/devart/commit/ff9164ce627998a886f2635b60cef5fde51fa7ac

**Comment by devjiwonchoi** [comment]
Cool! Just so you know, you can replace all those explicit `meta` tags with `generateMetadata`. 😄

**Comment by SH20RAJ** [comment]
> Cool! Just so you know, you can replace all those explicit `meta` tags with `generateMetadata`. 😄

I will do it 🥰

**Comment by SH20RAJ** [comment]
> Cool! Just so you know, you can replace all those explicit `meta` tags with `generateMetadata`. 😄

But what about client components bro 🤔

```
  × You are attempting to export "generateMetadata" from a component marked with "use client", which is disallowed. Either remove the export, or the "use client" directive. Read more: https://

```

**Comment by devjiwonchoi** [comment]
There are several ways to do it; I'd try to split the Client Components as children of `page`. If the `page` needs to be a Client Component, I would create a `layout` on that dynamic route and use `generateMetadata`.

**Comment by SH20RAJ** [comment]
> There are several ways to do it; I'd try to split the Client Components as children of `page`. If the `page` needs to be a Client Component, I would create a `layout` on that dynamic route and use `generateMetadata`.

While using the layout of the main route the content of layout is applying to all the child routes [...slugs] and title/meta overriding happening

**Comment by alexdln** [comment]
> But what about client components bro 🤔

Try to always keep page files as server components. This will help maintain years of happy life!)

If the entire page is client-side - take out its content into a separate component and make it client-side:

```tsx
// page.ts
const HomePage = () => (
  <HomePageContent />
)

// home-page-content.ts
'use client';

const HomePageContent = () => (
  // ...
)
```
